### ISSUE TITLE
frost(sdpa): f16 THD K/V descriptors lack the FP8 packed-total clamp — uninitialized capacity tail poisons O via 0*NaN

### ISSUE THREAD

**Issue by vedaanta** [NVIDIA/cudnn-frontend#718]
### Summary

The FP8/MXFP8 THD forward kernels patch their K/V TMA descriptor extents to the packed total `cu_k[B]` device-side, so the last sequence's tile tail lands as exact zeros. **The f16/bf16 THD paths have no such clamp** — their extent is the host-derived token capacity. When a caller binds a buffer whose rows `[total, capacity)` were never initialized, those rows are inside the extent, get loaded, and — being masked, so `P == 0` — reach BMM2 as `0 * NaN`, wiping whole tiles.

`#706` made the capacity *safe* in the sense of never leaving the caller's own allocation, but "caller-owned" is weaker than "initialized". This issue is the remaining half.

### Repro (SM100, bf16, cuDNN 9.26.0.33, `CUDNN_FRONTEND_ENABLE_FROST_ENGINES=1`)

THD graph, `seq_lens = [200, 150, 47]` (total 397), K/V bound as `(CAP, H, D)` packed buffers whose `[397, CAP)` tail is filled with NaN. Only the tail fill differs between runs:

| binding | NaNs in O |
|---|---|
| `CAP = 640` (> total), NaN tail | **201,728 — 49.6% of O** |
| `CAP = 640`, zero tail | 0 |
| `CAP = 397` (== total, extent exact), NaN tail | **0** |

The third row is the control: with the extent exactly equal to the packed total the tail is TMA-unreachable and NaN there is harmless. The failure appears only when capacity exceeds the total — precisely the window the clamp closes. Total 397 is not a multiple of `TILE_N`, so the last sequence's tail tile steps past it.

Repro script: attached below.

### Why it matters

The exposure is a caller that over-allocates and leaves the tail unwritten. That is exactly the pattern we recommend for **CUDA-graph capture** — allocate at capacity, capture once, replay with different `cu_seqlens` — so the hazard sits on the path the zero-host-read design (#552) is meant to unlock. `torch.empty` persistent KV buffers in a serving engine are the realistic instance; they usually hold stale-but-finite activations, which is why this has not bitten yet, but nothing guarantees that.

Current callers in-tree are immune, which is why no test catches it: `cudnn::sdpa_fwd` passes exactly `(T, H, D)` so capacity equals the total, and `test_mhas_v2` fills its entire allocated view.

### Fix

Port the existing FP8 mechanism. `build_thd_meta_o_kv_descs_kernel` already does all of it — copies the base K/V TensorMaps into `o_desc_words` slots `n_batch+1` / `n_batch+2`, patches `GLOBAL_DIM ord=2` to `cu_k[B]` read from the metadata it just wrote, and releases GENERIC→TENSORMAP. The f16 SM100 kernels currently call the O-only variant `build_thd_meta_o_descs_kernel`, and they already consume runtime descriptors elsewhere (`tma_slice_runtime_desc` for O), so the plumbing exists.

Needed:
- `o_desc_slots = b + 3` for f16 (currently `b + 1`);
- swap to the `_o_kv_descs` setup variant;
- rebind the `tma_k` / `tma_v` closures under `CFG.THD_VARLEN`, as `prefill_d128_fp8_sm100.py:642` does.

Affected: `prefill_d128_f16_sm100.py`, `prefill_d512_f16_sm100.py` (both on `build_thd_meta_o_descs_kernel`), and `prefill_f16_sm120.py`, which uses the meta-only `build_thd_meta_kernel` and so needs the K/V slots introduced rather than switched.

The one cost to measure is the `fence.proxy.tensormap::generic.acquire` on the K/V load path. `tma_tensormap_acquire` exists to hoist it to once per CTA (pass `acquire=False` per tile) rather than per load.

### Follow-on: this would retire the derived capacity

`tensormap_replace` overwrites `GLOBAL_DIM`, so a clamped tensor stops depending on the host value entirely — immune to both over-claim (this issue) and under-claim (#613's truncation mode). Extending the same patch to **Q** from `cu_q[B]` (one more slot) would make Q/K/V extents all exact device-side; O is already per-sequence exact and LSE is predicated per row. At that point `_thd_capacity` degrades from "must be ≥ the total **and** ≤ what the caller owns" to just "≤ what the caller owns" — the only thing buffer geometry can honestly promise.

Q does not need the clamp for NaN-safety (it is the parallel dimension: a garbage Q row poisons only its own S/P/O row, and that row is never stored), but it does need it to remove the last host-capacity dependency. Q also loads once per tile rather than once per KV iteration, so its acquire fence is the cheaper one.

<details><summary>Repro script</summary>

```python
"""Probe: does the f16 THD path read the capacity TAIL of an over-allocated
K/V buffer?

The FP8/MXFP8 setup kernel patches K/V descriptor extents to cu_k[B] so
tile-tail loads past the packed total land as exact zeros. The f16 path has no
such clamp -- its extent is the host-derived capacity. If a caller binds a
buffer whose rows [total, capacity) were never initialized, those rows are
inside the extent, get loaded, and (being masked, P == 0) reach BMM2 as
0 * NaN.

Differential: same graph, same data, only the TAIL fill differs (NaN vs 0).
If O differs, the tail is being read.
"""
import math
import os
import sys

sys.path.insert(0, "/home/scratch.vagarwalla_gpu/cudnn-fe-613/test/python")
sys.path.insert(0, "/home/scratch.vagarwalla_gpu/cudnn-fe-613/test/python/sdpa/frost")

import torch
import cudnn
from test_sdpa_fwd_dsl_sm100 import _select_engine, engine_name

dev = "cuda"
H, d = 8, 128
dtype = torch.bfloat16
seq_lens = [200, 150, 47]
B, S_max, T = len(seq_lens), max(seq_lens), sum(seq_lens)
CAP = 640  # > T (397): rows [397, 640) are the uninitialized capacity tail
cu = [0]
for s_i in seq_lens:
    cu.append(cu[-1] + s_i)
scale = 1.0 / math.sqrt(d)

torch.manual_seed(0)
q_pk = torch.randn(T, H, d, device=dev, dtype=dtype)
k_live = torch.randn(T, H, d, device=dev, dtype=dtype)
v_live = torch.randn(T, H, d, device=dev, dtype=dtype)


def run(tail_value, cap):
    """Bind K/V as (cap, H, d) buffers whose [T, cap) tail is `tail_value`."""
    k_buf = torch.full((cap, H, d), tail_value, device=dev, dtype=dtype)
    v_buf = torch.full((cap, H, d), tail_value, device=dev, dtype=dtype)
    k_buf[:T] = k_live
    v_buf[:T] = v_live

    stride = (S_max * H * d, d, H * d, 1)
    io = cudnn.data_type.BFLOAT16
    g = cudnn.pygraph(io_data_type=io, intermediate_data_type=cudnn.data_type.FLOAT, compute_data_type=cudnn.data_type.FLOAT)
    tq = g.tensor(dim=[B, H, S_max, d], stride=list(stride), data_type=io, name="q")
    tk = g.tensor(dim=[B, H, S_max, d], stride=list(stride), data_type=io, name="k")
    tv = g.tensor(dim=[B, H, S_max, d], stride=list(stride), data_type=io, name="v")
    slq = torch.tensor(seq_lens, dtype=torch.int32, device=dev).view(B, 1, 1, 1)
    cu_t = torch.tensor(cu, dtype=torch.int64, device=dev)
    ro = (cu_t * H * d).view(B + 1, 1, 1, 1)
    sq, skv = g.tensor_like(slq), g.tensor_like(slq)
    qro, kro, vro, oro = (g.tensor_like(ro) for _ in range(4))
    tq.set_ragged_offset(qro)
    tk.set_ragged_offset(kro)
    tv.set_ragged_offset(vro)
    o, _ = g.sdpa(
        name="sdpa", q=tq, k=tk, v=tv, generate_stats=False, attn_scale=scale,
        use_causal_mask=True, use_padding_mask=True, seq_len_q=sq, seq_len_kv=skv,
    )
    o.set_output(True).set_dim([B, H, S_max, d]).set_stride(list(stride))
    o.set_ragged_offset(oro)
    g.validate()
    g.build_operation_graph()
    g.create_execution_plans([cudnn.heur_mode.A])
    _select_engine(g, engine_name())
    g.check_support()
    g.build_plans()
    o_buf = torch.zeros(T, H, d, device=dev, dtype=dtype)
    vp = {tq: q_pk, tk: k_buf, tv: v_buf, o: o_buf, sq: slq, skv: slq, qro: ro, kro: ro, vro: ro, oro: ro}
    g.execute(vp, torch.empty(max(g.get_workspace_size(), 1), device=dev, dtype=torch.uint8))
    torch.cuda.synchronize()
    return o_buf


print(f"seq_lens={seq_lens} T={T} CAP={CAP}  (tail rows [{T}, {CAP}) = uninitialized)")
o_zero = run(0.0, CAP)
o_nan = run(float("nan"), CAP)
print(f"  zero-tail: NaNs in O = {torch.isnan(o_zero).sum().item()}")
print(f"  NaN-tail : NaNs in O = {torch.isnan(o_nan).sum().item()}  ({100.0*torch.isnan(o_nan).float().mean().item():.1f}%)")
same = torch.equal(o_zero, o_nan)
print(f"  identical: {same}")

# Control: capacity == total means the extent is exact -> tail unreachable.
o_zero_tight = run(0.0, T)
o_nan_tight = run(float("nan"), T)
print(f"tight CAP == T == {T}:")
print(f"  NaN-tail : NaNs in O = {torch.isnan(o_nan_tight).sum().item()}")
print(f"  identical to zero-tail: {torch.equal(o_zero_tight, o_nan_tight)}")

print()
if not same and torch.isnan(o_nan).any():
    print("CONFIRMED: the f16 THD path reads the capacity tail; an uninitialized")
    print("tail poisons O through 0 * NaN. The FP8 clamp would prevent this.")
else:
    print("NOT reproduced: the f16 tail is not reaching BMM2 for this shape.")
```
</details>

**Comment by vedaanta** [comment]
Duplicate of #624, which describes this same defect (THD zero-host-read binding K/V to buffer capacity; NaN in the capacity tail x masked P=0 poisoning P@V) and already has harness hardening in flight as #646.

Closing. The parts that were new here — the FP8 descriptor-clamp fix that already exists in-tree, the differential repro with the exact-extent control, the contracted-vs-parallel scoping of which tensors need it, and the padded-THD escalation — are consolidated into https://github.com/NVIDIA/cudnn-frontend/issues/624#issuecomment-5414777119. The repro script stays available in this issue's body.
